### ISSUE TITLE
Predictable json schema

### ISSUE THREAD

**Issue by pchalamet** [minamijoyo/tfschema#4]
Hello,

I'm using your tools as mean to extract plugins schemas but since json output depends largely on named attributes and block_types it's harder to parse.

I've reworked the code (can be found here: https://github.com/pchalamet/tfschema) in order to output this json format to be easily machine parseable. Here is an output example for `tfschema resource show -format=json azurerm_virtual_network`

```
{
    "attributes": [
        {
            "name": "address_space",
            "type": "List(String)",
            "required": true,
            "optional": false,
            "computed": false,
            "sensitive": false
        },
        {
            "name": "dns_servers",
            "type": "List(String)",
            "required": false,
            "optional": true,
            "computed": false,
            "sensitive": false
        },
        ...
    ],
    "block_types": [
        {
            "name": "subnet",
            "attributes": [
                {
                    "name": "address_prefix",
                    "type": "String",
                    "required": true,
                    "optional": false,
                    "computed": false,
                    "sensitive": false
                },
                {
                    "name": "security_group",
                    "type": "String",
                    "required": false,
                    "optional": true,
                    "computed": false,
                    "sensitive": false
                },
                ...
            ],
            "block_types": [],
            "nesting": 3,
            "min_items": 0,
            "max_items": 0
        }
    ]
}
```

This means attributes have no more children of unknown keys - it's an array of well defined structure (new field is `name`). Same for block_types.

Before submitting a PR, I would like to be sure you agree with aforementioned changed. If you agree, I will send you the PR.

Thanks btw for your cool work !

**Comment by minamijoyo** [comment]
@pchalamet Hi, thank you for opening the issue!

I can understand that it is hard to parse if the key of JSON is not predictable.

The implementation of tfscehma largely depends on the terraform core's type and terraform uses map in the `configschema.Block`.

https://github.com/hashicorp/terraform/blob/bb37637a139151be0ff9bf00c85c448538e5b3d2/config/configschema/schema.go#L16-L24

The 0.12-dev branch of hashicorp/terraform, which is next major release, moved `config.configschema` package to `configs.configschema`  (https://github.com/hashicorp/terraform/pull/18398), and added a grpc provider implementation (https://github.com/hashicorp/terraform/pull/18638).

Although the grpc proto schema uses `repeated` type, that is array, the implementation of `configschema.Block` seems to keep using a map.
If my understanding is correct, we need to continue using `configschema` to support both current go-plugin and new grpc providers.

Terraform v0.12 support of tfschema has not started yet, but in order to make it easy to follow future implementation changes of terraform, I would like to keep the same data structure as terraform if possible.

I'm glad your proposal, but it is difficult to accept yours as it is.
Is it possible to parse as a map with arbitrary keys in your use case? Let me hear your use case.

Thanks!

**Comment by pchalamet** [comment]
Hello,

In my fork, I just changed `block.go` and `nested_block.go` to reach this goal (changes here: https://github.com/minamijoyo/tfschema/compare/master...pchalamet:master)

Some notes about changes in my fork:
* fix invalid go imports (eg: `github.com/minamijoyo/tfschema/` -> `tfschema/`)
* renamed tfschema folder to plugins (build output clash using `go build` - worth defining a makefile btw)
* implement aforementioned changed in `block.go` & `nested_block.go`

Anyway, it's implemented with *only* changes to your sources. I can work cleaning up a PR if you agree (and splitting my changes in several PRs if needed).

Let me know because, really, it's really hard to parse Json without well defined schema ! I'm using your tool in my own tf DSL (here: https://github.com/pchalamet/NTerraform/blob/master/NTerraformGen/TerraformSchema.fs) and it's so dead easy to process now.

Thanks!

**Comment by minamijoyo** [comment]
Woah, are you building a Terraform DSL with C# and F#? It's cool 😄 

> renamed tfschema folder to plugins (build output clash using go build - worth defining a makefile btw)

Sorry for the inconvenience, I usually use `go build -o bin/tfschema`, and I should create a Makefile. I'll add it later. 

> implement aforementioned changed in block.go & nested_block.go

I have some concerns to worry about:

(1) If the data structure of terraform's core is changed in the future,  it is hard to follow up. Every time we fix it, we need to consider valid and consistent policies, and there is no official policy there.  

(2) If we define a new structure and a key name of JSON ourselves, duplication and inconsistency may occur in the future.

(3) I'm planning to implement a language server for the terraform using tfschema as a library and need to look up a type definition of a specific resource type fast, but using a slice is inefficient.

My proposal as follows:

As for (1) and (3), we should only adjust the JSON output by adding a new type for rendering, leaving the original data structure as a map.
In other words, extract format related code from the Block's instance method to a new package (e.g., printer), define a new type for output and call it from command package. 

As for (2), I think that using the same key name as the schema of the grpc provider implementation in terraform v0.12 would reduce the possibility of name conflicts.
That is, the name in the NestedBlock should use `type_name` instead of `name`.

What do you think?

**Comment by pchalamet** [comment]
Yeah that's basically what I'm prototyping :-)

I doubt terraform internal will dramatically changes (at least not something that could be adapt really easily) since a lot of plugins are published and most plugins are not even up-to-date with current model (random for eg).

But I can understand your concern & decoupling with plugins is welcomed. But I thought Block & NestedBlock structures were playing this role (talking about Block/NestedNlock and not schema.Block/schema.NestedBlock which are terraform classes).

Functions NewBlock() & NewNestedBlock() already provide the decoupling you are talking about - so I do not see why something else should be introduced. If terraform implementation changes, only these 2 functions (+ NewAttributes()) will need to be updated (it's basically the mapping between terraform to tfschema).

My guess is that you have already put everything in place to avoid strong coupling with terraform plugin model - it's used both for json and table rendering. Just adapt those 2 structures (NewBlock & NestedBlock) and this should be ok. At least this is my understanding of your code.

For naming - it's the hardest part ;-) - no problem about the name until it's well defined and meaningful - but `type_name` is probably a better name for sure.

**Comment by minamijoyo** [comment]
> Functions NewBlock() & NewNestedBlock() already provide the decoupling you are talking about - so I do not see why something else should be introduced.

Thanks, my thoughts became clear by explaining to you.
Probably I need to explain a bit more context about the background and what I want to do now on.

Looking at the current code base, it is natural that you wonder.
Because, when I first made tfschema.Block and NestedBlock, I intended to add a layer to customize the output exactly as you say.

On the other hand, what I learned by making tfschema and watching terraform commits is:
Using terraform as a Go library (not CLI) is very fragile.
Backward compatibility is not guaranteed even if functions and structures are exported.

At the time of writing this, here is the difference of the terraform core between the current stable v0.11.7 and the next major release development branch v0.12-dev. 

https://github.com/hashicorp/terraform/compare/v0.11.7...v0.12-dev

It's an awesome big bang release, Yeah!

I don't believe terraform internal is stable. Third-party tools using a terraform as a Go library should be carefully designed to continue to maintain.

And now I'm trying to implement a language server for Terraform which provides editor-independent autocompletion using tfschema as a Go library. (I don't know when it will be completed, it's also my hobby project in my free time.)

From my point of view, I feel changing the internal data structure of tfschema.Block and NestedBlock from map to slice for the convenience of the only JSON format as CLI is a wrong direction, and the data model of tfschema should be separated from the requirements of the output format if I use tfschema as a Go library.

I can understand that you feel it's an excessive abstraction from the current code base as CLI, so if you can not agree, I will fix the code by myself to change the JSON format.

**Comment by minamijoyo** [comment]
I'll let you know when I implement a well defined easy-to-parse JSON schema.

**Comment by minamijoyo** [comment]
@pchalamet Fixed in #6.
Is there any problem with this output format?

**Comment by pchalamet** [comment]
Answered on the PR instead... This is perfect!
